File: py38/sort/heap.py
from utils import timer

from typing import List, Optional, Any


class SmallHeap:

    def __init__(self, l: List):
        self._l = l
        self.make_heap()

    def sift_up(self, index: int):
        # 运算的下标比实际的下标大 1
        j = index + 1
        while j >= 2:
            i = j // 2
            if self._l[j-1] < self._l[i-1]:
                self._swap(j-1, i-1)
            else:
                break

            j //= 2

    def sift_down(self, index: int):
        # 运算的下标比实际的下标大1
        j = index + 1
        while j * 2 <= len(self._l):
            i = j * 2
            if i < len(self._l) and self._l[i] < self._l[i-1]:
                i += 1
            if self._l[i-1] < self._l[j-1]:
                self._swap(j-1, i-1)
                j = i
            else:
                break

    def make_heap(self):
        # 对所有的非叶子节点进行 sift_down
        for i in range(len(self._l) // 2, -1, -1):
            self.sift_down(i)

    def pop_top(self) -> Optional[Any]:
        if len(self._l) > 1:
            top_element = self._l[0]
            self._l[0] = self._l.pop()
            self.sift_down(0)
            return top_element

        elif len(self._l) > 0:
            return self._l.pop()

        return None

    def _swap(self, index1, index2):
        self._l[index1], self._l[index2] = self._l[index2], self._l[index1]


@timer
def heap_sort(l: List) -> List:
    small_heap = SmallHeap(l)

    sorted_list = []

    while (top_element := small_heap.pop_top()) is not None:
        sorted_list.append(top_element)

    return sorted_list



if __name__ == '__main__':
    import random
    l = list(range(500000, 0, -1))
    random.shuffle(l)
    assert heap_sort(l) == list(range(1, 500001))

File: rs/src/sort/heap.rs
pub fn heap_sort<T: Ord + Copy>(l: Vec<T>) -> Vec<T>{
    let mut sorted_list: Vec<T> = Vec::with_capacity(l.capacity());

    let mut small_heap = SmallHeap::new(l);

    while !small_heap.is_empty() {
        sorted_list.push(small_heap.pop_top());
    }

    sorted_list
}

pub struct SmallHeap<T: Ord + Copy> {
    list: Vec<T>,
}

impl<T: Ord + Copy> SmallHeap<T> {
    pub fn new(list: Vec<T>) -> SmallHeap<T> {
        let mut small_heap = SmallHeap {
            list
        };

        small_heap.make_heap();
        small_heap
    }

    pub fn sift_down(&mut self, index: usize) {
        let mut j = index + 1;

        while 2 * j <= self.list.len() {
            let mut i = 2 * j;
            if i < self.list.len() && self.list[i] < self.list[i-1] {
                i += 1;
            }
            if self.list[i-1] < self.list[j-1] {
                self.list.swap(i-1, j-1);
                j = i;
            }
            else {
                break;
            }
        }
    }

    fn make_heap(&mut self) {
        for i in (0..self.list.len() / 2).rev() {
            self.sift_down(i);
        }
    }

    pub fn pop_top(&mut self) -> T {
        match self.list.len() {
            0 => panic!("list is empty!"),
            1 => self.list.pop().unwrap(),
            _ => {
                let top_element = self.list[0];
                self.list[0] = self.list.pop().unwrap();
                self.sift_down(0);
                top_element
            }
        }
    }

    pub fn is_empty(&self) -> bool {
        self.list.len() == 0
    }
}




